mixer/adapter/kubernetesenv/kubernetesenv.go

```diff
@@ -123,32 +129,122 @@ func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler,
 	// provide this basic functionality before.
 	b.Lock()
 	defer b.Unlock()
-	controller, found := b.controllers[path]
+	_, found := b.controllers[path]
 	if !found {
 		clientset, err := b.newClientFn(path, env)
 		if err != nil {
 			return nil, fmt.Errorf("could not build kubernetes client: %v", err)
 		}
-		controller = newCacheController(clientset, refresh, env)
-		env.ScheduleDaemon(func() { controller.Run(stopChan) })
-		// ensure that any request is only handled after
-		// a sync has occurred
-		env.Logger().Infof("Waiting for kubernetes cache sync...")
-		if success := cache.WaitForCacheSync(stopChan, controller.HasSynced); !success {
-			stopChan <- struct{}{}
-			return nil, errors.New("cache sync failure")
+
+		controller, err = getNewCacheController(b, clientset, env)
+		if err != nil {
+			return nil, fmt.Errorf("could not create new cache controller: %v", err)
 		}
-		env.Logger().Infof("Cache sync successful.")
+
+		controllers = append(controllers, controller)
 		b.controllers[path] = controller
+
+		remoteControllers, err = createRemoteCacheControllers(b, clientset, env)
+		if err == nil {
+			if len(remoteControllers) != 0 {
+				controllers = append(controllers, remoteControllers...)
+			}
+		} else {
+			return nil, fmt.Errorf("failure on creating remote controllers: %v", err)
+		}
+	} else {
+		for key := range b.controllers {
+			controllers = append(controllers, b.controllers[key])
+		}
 	}
 
+	env.Logger().Infof("installed %d controllers", len(controllers))
+
 	return &handler{
 		env:      env,
-		k8sCache: controller,
+		k8sCache: controllers,
 		params:   paramsProto,
 	}, nil
 }
 
+func createRemoteCacheControllers(b *builder, clientset k8s.Interface,
+	env adapter.Env) ([]cacheController, error) {
+	var controller cacheController
+	var k8sInterface k8s.Interface
+	var remoteControllers []cacheController
+	var opts meta_v1.ListOptions
+
+	secretNamespace := "istio-system" // TODO: make configurable
```

Comment:
can we add these values to the config proto with default values?  That should be straightforward.

Reply:
Yes. I started looking into that when the overall discussion of how mixer should be deployed in a multi-cluster environment and conversations of perhaps creating a common library of the secrets code that can be consumed by pilot and mixer (and future) came up. Just punted for now.